src/ui/map.js

```diff
@@ -1185,45 +1190,52 @@ class Map extends Camera {
         } else if (typeof projection === 'string') {
             projection = (({name: projection}: any): ProjectionSpecification);
         }
-        return this._updateProjection(projection);
+        this._explicitProjection = projection;
+        return this._updateProjection();
     }
 
-    _updateProjection(explicitProjection?: ProjectionSpecification | null): this {
-        const prevProjection = this.getProjection();
-        if (explicitProjection === null) {
-            this._explicitProjection = null;
+    _updateProjectionTransition() {
+        // The projection isn't globe, we can skip updating the transition
+        if (this.getProjection().name !== 'globe') {
+            return;
         }
-        const projection = explicitProjection || this.getProjection();
 
+        const tr = this.transform;
+        const projection = tr.projection.name;
         let projectionHasChanged;
-        // At high zoom on globe, set transform projection to Mercator while _explicitProjection stays globe.
-        if (projection && projection.name === 'globe' && this.transform.zoom >= GLOBE_ZOOM_THRESHOLD_MAX) {
-            projectionHasChanged = this.transform.setProjection({name: 'mercator'});
-            this.transform.mercatorFromTransition = true;
-        } else {
-            projectionHasChanged = this.transform.setProjection(projection);
-            this.transform.mercatorFromTransition = false;
+
+        if (projection === 'globe' && tr.zoom >= GLOBE_ZOOM_THRESHOLD_MAX) {
+            tr.setMercatorFromTransition();
+            projectionHasChanged = true;
+        } else if (projection === 'mercator' && tr.zoom < GLOBE_ZOOM_THRESHOLD_MAX) {
+            projectionHasChanged = tr.setProjection({name: 'globe'});
         }
 
-        // When called through setProjection, update _explicitProjection
-        if (explicitProjection) {
-            this._explicitProjection = (explicitProjection.name === "globe" ?
-                {name:'globe', center:[0, 0]} :
-                this.transform.getProjection());
+        if (projectionHasChanged) {
+            this.style.applyProjectionUpdate();
+            this.style._forceSymbolLayerUpdate();
+        }
+    }
+
+    _updateProjection(): this {
+        // Check explicit projection separately as transform.projection hasn't been updated yet
+        const projection = this._explicitProjection || this.getProjection();
```

Comment:
As above, so long as `_explicitProjection` is a projection object, can we place this logic in `getProjection` and simplify this to:
```suggestion
        const projection = this.getProjection();
```